### ISSUE TITLE
Active failover fails to install foxx app

### ISSUE THREAD

**Issue by rivet351** [arangodb/arangodb#13915]
## My Environment

* __ArangoDB Version__:        3.7.6
* __Storage Engine__:         RocksDB
* __Deployment Mode__:        Active Failover
* __Deployment Strategy__:    Kubernetes
* __Configuration__:          Azure Kubernetes with load balancer
* __Infrastructure__:         Azure
* __Operating System__:       Azure Kubernetes with docker
* __Total RAM in your machine__: 32gb
* __Disks in use__:        	SSD 
* __Used Package__:             Docker - official Docker library 

## Component, Query & Data

__Affected feature__:
Installation of Foxx app following failover event

__Replication Factor & Number of Shards (Cluster only)__:
Leader with 2 x Followers

## Steps to reproduce

1. Kill leader to trigger failover
2. Inspect UI (Foxx app sometimes appears, sometimes does not and is intermittent on refresh)
3. Error seen in UI and kubernetes logs:

2021-04-06T15:18:32Z [1] ERROR [24213] Failed to load Foxx service mounted at "/mailapi"
2021-04-06T15:18:32Z [1] ERROR [24213] via ArangoError: service files missing
2021-04-06T15:18:32Z [1] ERROR [24213] Mount: /mailapi
2021-04-06T15:18:32Z [1] ERROR [24213]     at loadInstalledService (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/manager.js:616:13)
2021-04-06T15:18:32Z [1] ERROR [24213]     at initLocalServiceMap (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/manager.js:519:23)
2021-04-06T15:18:32Z [1] ERROR [24213]     at selfHeal (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/manager.js:245:5)
2021-04-06T15:18:32Z [1] ERROR [24213]     at Object.selfHealAll [as healAll] (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/manager.js:196:20)
2021-04-06T15:18:32Z [1] ERROR [24213]     at Object.exports.manage (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/queues/manager.js:234:19)
2021-04-06T15:18:32Z [1] ERROR [24213]     at eval (eval at <anonymous> (unknown source), <anonymous>:2:50)
2021-04-06T15:18:32Z [1] ERROR [24213]     at eval (eval at <anonymous> (unknown source), <anonymous>:3:9)
2021-04-06T15:18:32Z [1] ERROR [24213]     at eval (eval at <anonymous> (unknown source), <anonymous>:3:21)

__Problem__:
The failover does look to take place (new leader is elected and the service attempts to install the foxx app previously working on the original leader. However we see error 503 - foxx app sometimes viewable, sometimes not - the foxx app then needs hard deleting before a new installation working installation can be done. Until this is done the foxx app remains in a partially broken state (sometimes performing the jobs as expected, otherwise returning 503 errors)

__Expected result__:
Foxx app reinstalls without errors on new leader following failover

**Comment by Simran-B** [comment]
Hi @rivet351,
3.7.9 has a fix related to Foxx self-heal. It doesn't seem to be related to your issue, but could you try the latest version anyway?

Are you able to reproduce the problem in a local setup or only on Azure? I wonder if it's Azure specific or not.

Does the initial leader come back up as follower and remain follower?

Can you compare the contents of the `_appbundles` system collections and the Foxx service files in the file system of both leader and follower? I wonder if there's a discrepancy at some point. Also, what is the file system? Maybe some files are actually not available temporarily for some reason?

**Comment by rivet351** [comment]
Hi Simran,

We upgraded to 3.7.10 and have repeated multiple times with failover completing successfully so it seems the self-heal was the issue. Thanks for your help and I'm closing the ticket.

**Comment by rivet351** [comment]
Hi, 

I've re-opened the ticket as we have now had this happen to our leader without a failover event (same errors as seen in the stack trace above).

**Comment by pluma** [comment]
@rivet351 Can you confirm that this happened after an upgrade/replace of the Foxx app on the leader which then/later produced this log output?

**Comment by rivet351** [comment]
Hi,

It looks to have happened randomly, several days after our last deployment. Timeline for this looks like:
05/05 - deployed a new Foxx app version to the leader
06/05 - 09/05 - worked as normal
10/05 - error above seen - no failover - required manually deleting and triggering a re-install to work again

**Comment by cubeover** [comment]
we too are seeing the issue on 3.7.6.

**Comment by cubeover** [comment]
```
 [24213] at eval (eval at (unknown source), :3:21)
 [24213] at eval (eval at (unknown source), :3:9)
[24213] at eval (eval at (unknown source), :2:50)
[24213] at Object.exports.manage (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/queues/manager.js:222:19)
24213] at Object.selfHealAll [as healAll] (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/manager.js:192:20)
[4213] at selfHeal (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/manager.js:231:5)
[24213] at initLocalServiceMap (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/manager.js:439:23)
[24213] at loadInstalledService (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/manager.js:519:13)
[24213] Mount: /foxx
[24213] via ArangoError: service files outdated
[24213] Failed to load Foxx service mounted at "/foxx"
[24213] at eval (eval at (unknown source), :3:21)
[24213] at eval (eval at (unknown source), :3:9)
[24213] at eval (eval at (unknown source), :2:50)
[24213] at Object.exports.manage (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/queues/manager.js:222:19)
[24213] at Object.selfHealAll [as healAll] (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/manager.js:192:20)
[24213] at selfHeal (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/manager.js:231:5)
[24213] at initLocalServiceMap (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/manager.js:439:23)
[24213] at loadInstalledService (/volumes/data/ARANGO/resilientsingle8529/data/js/server/modules/@arangodb/foxx/manager.js:519:13)
[24213] Mount: /foxx
```

**Comment by rivet351** [comment]
Hi,

Any update on this?

**Comment by Simran-B** [comment]
Unfortunately not. I created an internal ticket https://arangodb.atlassian.net/browse/BTS-484 for tracking the issue.

**Comment by dothebart** [comment]
Hi, 
This has been fixed by this PR: https://github.com/arangodb/arangodb/pull/14754 which is available for download with the latest ArangoDB 3.7 releases.
I'm sorry we didn't comment this here too since the missing self heal was discovered by internal tests independently as well and hence fixed without communicating here.
